### DISCUSSION TITLE
How do I ensure that server communication is finished before ending my experiment?

### DISCUSSION THREAD

**Discussion by sophief10** [jspsych/jsPsych#2320] [Q&A] [answered]
Hello everyone!

I am trying to integrate a jsPsych experiment with Prolific, and am following the [tutorial](https://www.jspsych.org/7.0/overview/prolific/#completing-the-experiment) on the jsPsych website. 

The tutorial recommends that I ensure the completion of all server communication before ending the experiment, and links me to this example:

```
var saveData = {
            type: 'call-function',
            async: true,
            func: function(done) {
                var xhttp = new XMLHttpRequest();
                xhttp.onreadystatechange = function() {
                    if (this.readState == 4 && this.status == 200) {
                        var response_data = xhttp.responseText;
                        done(response_data);
                    }
                };
                xhttp.open("GET", "path_to_server_script.php", true);
                xhttp.send();
            }  
        };
```

I do not know XML so I have no idea what is happening in the above code. I copied it exactly into my own code, but it does not seem to work because my browser will not show the slide that comes after the XML function trial. 

Does anyone know how to adjust the jsPsych.org example to fit their own experiment? Or, would someone please explain what is happening in the XML code above so that I can adjust it? 

Here is the rest of my code for reference:

```

        //EXPERIMENT SET-UP

        var finalTimeline = []; 

        //capture info from Prolific

        var subject_id = jsPsych.data.getURLVariable('PROLIFIC_PID');
        var study_id = jsPsych.data.getURLVariable('STUDY_ID');
        var session_id = jsPsych.data.getURLVariable('SESSION_ID');

        jsPsych.data.addProperties({
            subject_id: subject_id,
            study_id: study_id,
            session_id: session_id
        });


        //ORIENTATION

        var welcome = {
            type: 'html-keyboard-response',
            stimulus: 'Welcome to our experiment! Press any key to begin.'
        };
        finalTimeline.push(welcome);



        //END EXPERIMENT

        //ensure all data are saved

        var saveData = {
            type: 'call-function',
            async: true,
            func: function(done) {
                var xhttp = new XMLHttpRequest();
                xhttp.onreadystatechange = function() {
                    if (this.readState == 4 && this.status == 200) {
                        var response_data = xhttp.responseText;
                        done(response_data);
                    }
                };
                xhttp.open("GET", "path_to_server_script.php", true);
                xhttp.send();
            }  
        };
        finalTimeline.push(saveData);

        var test = {
            type: 'html-keyboard-response',
            stimulus: 'test'
        };
        finalTimeline.push(test);


        //RUN EXPERIMENT

        jsPsych.init({
            timeline: finalTimeline,
            show_progress_bar: true
        });
```

**Comment by becky-gilbert** [comment, marked as answer]
Hi @woodenchair, what kind of server/platform are you using to host your experiment? The example code from the docs assumes that you're using php, and that you have a script called "path_to_server_script.php", so that might be the problem. Do you see any errors int he JavaScript console when your experiment gets to the `saveData` trial?

To answer your question more generally: the main idea with ensuring that your data is saved before the experiment ends is to send the data and use a callback function or promise so that the experiment only continues after the server responds (i.e. confirms that it has received the data). This is why the way that you do this depends on your server. 

With the php example, the `call-function` trial is sending the data (xhttp.send) and then only ending the trial when the server response is received (checked via the if statement inside the onreadystatechanged function). When this happens, the `done` function is called, which is what ends the jsPsych trial and continues the experiment.

With cognition.run, the data is transferred after every trial, so you shouldn't need to worry about handling this. But you might want to read this thread, and Javier's comment in particular, about saving the data when you also want to redirect a participant at the end of the experiment: https://github.com/jspsych/jsPsych/discussions/1850#discussioncomment-1142537

With other server platforms and tools, there might be a "send data" function that you can use to send the data and then do something when the the data transfer is successful. This is might be a parameter called something like "on_success callback function", and you can use that to continue or end the experiment after the data transfer is finished.

**Reply by sophief10** [reply]
@becky-gilbert So it would be okay for me to simply remove the saveData trial and keep the rest of the experiment as is? I will be sure not the call endExperiment as per the advice in the thread you shared

**Reply by becky-gilbert** [reply]
Yes absolutely. The saveData trial is just one example of how you could save the data if you're hosting your study on a custom server with php.

**Reply by sophief10** [reply]
Great! Thank you, @becky-gilbert !
